processor/src/chiplets/mod.rs

```diff
@@ -151,8 +136,60 @@ impl Chiplets {
         path: &[Word],
         index: Felt,
     ) -> (Felt, Word, Word) {
-        self.hasher
-            .update_merkle_root(old_value, new_value, path, index)
+        let (addr, old_root, new_root, lookups) = self
+            .hasher
+            .update_merkle_root(old_value, new_value, path, index);
+        self.bus.request_hasher_operation(lookups, self.clk);
+
+        (addr, old_root, new_root)
+    }
+
+    // HASH CHIPLET ACCESSORS FOR CONTROL BLOCK DECODING
+    // --------------------------------------------------------------------------------------------
+
+    /// Requests the hash of the provided words from the Hash chiplet and returns the result
+    /// hash(h1, h2).
+    ///
+    /// The returned tuple also contains the row address of the execution trace at which hash
+    /// computation started.
+    pub fn merge_control_blocks(&mut self, h1: Word, h2: Word) -> (Felt, Word) {
```

Comment:
> Also, one thing that bothers me a little is that this and hash_span_block methods look identical to permute and other methods above, but they behave differently (e.g,, they actually enqueue the requests and results are read in different rows). I wonder if we should change the signatures to match this behavior.

I agree in general that something here isn't as good as it can be - it's why I regrouped the functions, added a new section header, and renamed them to include a reference to control blocks. However, I'm not sure that adjusting the signatures is the best option. For one thing, the other "non-control-block" functions have a variety of signatures. For another, some of these parameters could be unnecessary or redundant, as you noted.

Fundamentally the issue here is that all other functions in the chiplets are being called by operation executors operating on the Stack, which means that they're all treated in essentially the same way, whether they're requesting lookups from the Bitwise, Memory, or Hash chiplets. (The only difference with the Hasher is that an operation requires 2 or 4 lookups instead of 1).

These 3 functions (currently `merge_control_blocks`, `hash_span_block`, `request_control_block_hash`) are called by the decoder, and the needs are different.

I wonder if we want to have a different access point for the decoder instead. One possible approach would be to have a single decoder access point with a context that indicates if it's initializing, continuing (absorb), or ending a control block. Obviously they need different information, so I'm not sure yet what the best way to go about this is, but finding a good way to group this together and separate it from the other functions for "operation handling" could help make things clearer

Reply:
The other thing which is a bit less existential but relevant now is that `request_control_block_hash` currently covers both the proposed `read_hash_result` and `absorb_span_batch`, so if we were to make the specific proposed changes now, I think it would basically force the refactors that we've discussed doing later in a second pass (related to the queued lookups and refactoring functions inside the hasher itself.

I do think we want to revisit this, and it won't be as bad now that there are tests in place (although we should probably write a test for mrupdate first), but it's worth noting that I think it's nontrivial

Reply:
(Either that or we'd change to these signatures, leave TODOs, and not actually use the parameters, but just have functions with different signatures calling `self.bus.send_queued_hasher_request(self.clk)`, and maybe this is what you're proposing and I've misunderstood)

Reply:
I've made a minimal set of changes to improve on this and address other comments, but it doesn't close this question.